vpr/src/place/net_cost_handler.cpp

```diff
@@ -0,0 +1,2277 @@
+#include "net_cost_handler.h"
+#include "globals.h"
+#include "placer_globals.h"
+#include "move_utils.h"
+#include "place_timing_update.h"
+#include "noc_place_utils.h"
+#include "vtr_math.h"
+
+using std::max;
+using std::min;
+
+/* Flags for the states of the bounding box.                              *
+ * Stored as char for memory efficiency.                                  */
+
+enum class NetUpdateState {
+    NOT_UPDATED_YET,
+    UPDATED_ONCE,
+    GOT_FROM_SCRATCH
+};
+
+/* This defines the error tolerance for floating points variables used in *
+ * cost computation. 0.01 means that there is a 1% error tolerance.       */
+#define ERROR_TOL .01
+
+/* Expected crossing counts for nets with different #'s of pins.  From *
+ * ICCAD 94 pp. 690 - 695 (with linear interpolation applied by me).   *
+ * Multiplied to bounding box of a net to better estimate wire length  *
+ * for higher fanout nets. Each entry is the correction factor for the *
+ * fanout index-1                                                      */
+static const float cross_count[50] = {/* [0..49] */ 1.0, 1.0, 1.0, 1.0828,
+                                      1.1536, 1.2206, 1.2823, 1.3385, 1.3991, 1.4493, 1.4974, 1.5455, 1.5937,
+                                      1.6418, 1.6899, 1.7304, 1.7709, 1.8114, 1.8519, 1.8924, 1.9288, 1.9652,
+                                      2.0015, 2.0379, 2.0743, 2.1061, 2.1379, 2.1698, 2.2016, 2.2334, 2.2646,
+                                      2.2958, 2.3271, 2.3583, 2.3895, 2.4187, 2.4479, 2.4772, 2.5064, 2.5356,
+                                      2.5610, 2.5864, 2.6117, 2.6371, 2.6625, 2.6887, 2.7148, 2.7410, 2.7671,
+                                      2.7933};
+
+/* The arrays below are used to precompute the inverse of the average   *
+ * number of tracks per channel between [subhigh] and [sublow].  Access *
+ * them as chan?_place_cost_fac[subhigh][sublow].  They are used to     *
+ * speed up the computation of the cost function that takes the length  *
+ * of the net bounding box in each dimension, divided by the average    *
+ * number of tracks in that direction; for other cost functions they    *
+ * will never be used.                                                  *
+ */
```

Comment:
The vertical cost function is added in another branch. I will make a PR for that after this one. I made that change close to my defense, and because of that, I was not able to make a PR for it.